src/deploy/functions/prepare.ts

```diff
@@ -76,14 +84,85 @@ export async function prepare(
     runtimeConfig = { ...runtimeConfig, ...(await getFunctionsConfig(projectId)) };
   }
 
-  // ===Phase 1. Load codebase from source.
-  context.sources = {};
   const codebaseUsesEnvs: string[] = [];
   const wantBackends: Record<string, backend.Backend> = {};
+  // ===Phase 1. Load codebase from source.
+  await preparePhase1(
```

Comment:
the "preparePhase1" name makes me nervous because it relies on authors keeping the comments up to date, which is notoriously more difficult than keeping function/variable names up to date. Consider giving them slightly more obvious names.

Reply:
+1

It's hard to extract these logic out because it's requires a very specific execution ordering. I think you might have done this so you can more easily test out your code, but I'd recommend that we unit test individual pieces of logic that easily testable (e.g. generating hash for each environment) and intead rely on some form of integration testing to test the end-to-end workflow.

We can brainstorm how we can setup an integration test together if you want.

Reply:
+100

Naming based on functionality is way more useful than order, I totally agree. I didn't have the chance to fully read through functionality of each "phase" before coming up with a better name 🤷

Reply:
Reverted this bit, will revisit later.